### ISSUE TITLE
Crash during find

### ISSUE THREAD

**Issue by chrisridd** [Sigil-Ebook/Sigil#156]
I did a case-sensitive find in the current HTML file for "biD" and Sigil 0.9.0 crashed.

Other find options were: Dot All, Minimal Match, and Wrap.

I can provide the commercial epub (with DRM removed) privately if necessary - this doesn't happen in an empty book so it may be data-sensitive.

The stack trace from OS X is:

```
Process:               Sigil [24678]
Path:                  /Applications/Sigil.app/Contents/MacOS/Sigil
Identifier:            com.sigil-ebook.Sigil.app
Version:               0.9.0 (0.9.0)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Sigil [24678]
User ID:               501

Date/Time:             2015-11-15 13:59:39.757 +0000
OS Version:            Mac OS X 10.11.1 (15B42)
Report Version:        11
Anonymous UUID:        45C230B7-358D-FD17-0553-04F770253269

Sleep/Wake UUID:       6CE9819C-D92A-472E-BEF1-AE469E850C64

Time Awake Since Boot: 380000 seconds
Time Since Wake:       4700 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

VM Regions Near 0:
--> 
    __TEXT                 000000010f476000-000000010f962000 [ 5040K] r-x/rwx SM=COW  /Applications/Sigil.app/Contents/MacOS/Sigil

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   com.sigil-ebook.Sigil.app     	0x000000010f6ee834 BookViewPreview::GetSelectionOffset(QMap<int, QString> const&, Searchable::Direction) + 68
1   com.sigil-ebook.Sigil.app     	0x000000010f6ee04a BookViewPreview::FindNext(BookViewPreview::SearchTools&, QString const&, Searchable::Direction, bool, bool, bool, bool) + 314
2   com.sigil-ebook.Sigil.app     	0x000000010f6ed89d BookViewPreview::FindNext(QString const&, Searchable::Direction, bool, bool, bool, bool) + 93
3   com.sigil-ebook.Sigil.app     	0x000000010f7ab253 non-virtual thunk to BookViewEditor::FindNext(QString const&, Searchable::Direction, bool, bool, bool, bool) + 35
4   com.sigil-ebook.Sigil.app     	0x000000010f763fe9 FindReplace::FindText(Searchable::Direction) + 329
5   com.sigil-ebook.Sigil.app     	0x000000010f762920 FindReplace::Find() + 96
6   com.sigil-ebook.Sigil.app     	0x000000010f7628b1 FindReplace::FindClicked() + 33
7   org.qt-project.QtCore         	0x00000001130567d2 QMetaObject::activate(QObject*, int, int, void**) + 2994
8   org.qt-project.QtWidgets      	0x00000001101b9e40 QAbstractButton::clicked(bool) + 64
9   org.qt-project.QtWidgets      	0x000000010ff1373a 0x10fdfd000 + 1140538
10  org.qt-project.QtWidgets      	0x000000010ff135b7 0x10fdfd000 + 1140151
11  org.qt-project.QtWidgets      	0x000000010ff145be QAbstractButton::mouseReleaseEvent(QMouseEvent*) + 270
12  org.qt-project.QtWidgets      	0x000000010ffdaf7f QToolButton::mouseReleaseEvent(QMouseEvent*) + 15
13  org.qt-project.QtWidgets      	0x000000010fe6e24c QWidget::event(QEvent*) + 1628
14  org.qt-project.QtWidgets      	0x000000010ff14313 QAbstractButton::event(QEvent*) + 179
15  org.qt-project.QtWidgets      	0x000000010ffdb544 QToolButton::event(QEvent*) + 324
16  org.qt-project.QtWidgets      	0x000000010fe2eddb QApplicationPrivate::notify_helper(QObject*, QEvent*) + 251
17  org.qt-project.QtWidgets      	0x000000010fe3263d QApplication::notify(QObject*, QEvent*) + 9485
18  org.qt-project.QtCore         	0x0000000113023da3 QCoreApplication::notifyInternal(QObject*, QEvent*) + 115
19  org.qt-project.QtWidgets      	0x000000010fe2f75b QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) + 987
20  org.qt-project.QtWidgets      	0x000000010fe892a1 0x10fdfd000 + 574113
21  org.qt-project.QtWidgets      	0x000000010fe884df 0x10fdfd000 + 570591
22  org.qt-project.QtWidgets      	0x000000010fe2eddb QApplicationPrivate::notify_helper(QObject*, QEvent*) + 251
23  org.qt-project.QtWidgets      	0x000000010fe320f8 QApplication::notify(QObject*, QEvent*) + 8136
24  org.qt-project.QtCore         	0x0000000113023da3 QCoreApplication::notifyInternal(QObject*, QEvent*) + 115
25  org.qt-project.QtGui          	0x00000001128abc73 QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) + 2163
26  org.qt-project.QtGui          	0x00000001128aab13 QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) + 131
27  org.qt-project.QtGui          	0x000000011289844a QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 346
28  libqcocoa.dylib               	0x0000000115d3a4d1 0x115d19000 + 136401
29  com.apple.CoreFoundation      	0x00007fff8f3678b1 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
30  com.apple.CoreFoundation      	0x00007fff8f3470ac __CFRunLoopDoSources0 + 556
31  com.apple.CoreFoundation      	0x00007fff8f3465cf __CFRunLoopRun + 927
32  com.apple.CoreFoundation      	0x00007fff8f345fc8 CFRunLoopRunSpecific + 296
33  com.apple.HIToolbox           	0x00007fff9b678d55 RunCurrentEventLoopInMode + 235
34  com.apple.HIToolbox           	0x00007fff9b678a97 ReceiveNextEventCommon + 184
35  com.apple.HIToolbox           	0x00007fff9b6789cf _BlockUntilNextEventMatchingListInModeWithFilter + 71
36  com.apple.AppKit              	0x00007fff8d5acd96 _DPSNextEvent + 1067
37  com.apple.AppKit              	0x00007fff8d5ac1c5 -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454
38  com.apple.AppKit              	0x00007fff8d5a0d28 -[NSApplication run] + 682
39  libqcocoa.dylib               	0x0000000115d393bd 0x115d19000 + 132029
40  org.qt-project.QtCore         	0x000000011302139d QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 381
41  org.qt-project.QtCore         	0x000000011302437a QCoreApplication::exec() + 346
42  com.sigil-ebook.Sigil.app     	0x000000010f47d3cb main + 4667
43  libdyld.dylib                 	0x00007fff8a8cd5ad start + 1

Thread 1:: Dispatch queue: com.apple.libdispatch-manager
0   libsystem_kernel.dylib        	0x00007fff95f0d0a2 kevent_qos + 10
1   libdispatch.dylib             	0x00007fff9997f11d _dispatch_mgr_invoke + 216
2   libdispatch.dylib             	0x00007fff9997ed85 _dispatch_mgr_thread + 52

Thread 2:
0   libsystem_kernel.dylib        	0x00007fff95f0c78a __workq_kernreturn + 10
1   libsystem_pthread.dylib       	0x00007fff9098e58c _pthread_wqthread + 1283
2   libsystem_pthread.dylib       	0x00007fff9098c375 start_wqthread + 13

Thread 3:
0   libsystem_kernel.dylib        	0x00007fff95f0c78a __workq_kernreturn + 10
1   libsystem_pthread.dylib       	0x00007fff9098e58c _pthread_wqthread + 1283
2   libsystem_pthread.dylib       	0x00007fff9098c375 start_wqthread + 13

Thread 4:
0   libsystem_kernel.dylib        	0x00007fff95f0c78a __workq_kernreturn + 10
1   libsystem_pthread.dylib       	0x00007fff9098e58c _pthread_wqthread + 1283
2   libsystem_pthread.dylib       	0x00007fff9098c375 start_wqthread + 13

Thread 5:: Qt bearer thread
0   libsystem_kernel.dylib        	0x00007fff95f0c222 __select + 10
1   libsystem_c.dylib             	0x00007fff94c8e896 pselect$1050 + 181
2   org.qt-project.QtCore         	0x0000000113074ce8 qt_safe_select(int, fd_set*, fd_set*, fd_set*, timespec const*) + 104
3   org.qt-project.QtCore         	0x0000000113075da0 QEventDispatcherUNIXPrivate::doSelect(QFlags<QEventLoop::ProcessEventsFlag>, timespec*) + 672
4   org.qt-project.QtCore         	0x0000000113076e6a QEventDispatcherUNIX::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 234
5   org.qt-project.QtCore         	0x000000011302139d QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 381
6   org.qt-project.QtCore         	0x0000000112e16a95 QThread::exec() + 117
7   org.qt-project.QtCore         	0x0000000112e1a523 0x112de5000 + 218403
8   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
9   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
10  libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 6:: JavaScriptCore::BlockFree
0   libsystem_kernel.dylib        	0x00007fff95f0bf5e __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff9098f73d _pthread_cond_wait + 767
2   org.qt-project.QtWebKit       	0x0000000111acd0f8 0x11099d000 + 18022648
3   org.qt-project.QtWebKit       	0x0000000111dbc74f 0x11099d000 + 21100367
4   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
5   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
6   libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 7:: JavaScriptCore::Marking
0   libsystem_kernel.dylib        	0x00007fff95f0bf5e __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff9098f73d _pthread_cond_wait + 767
2   org.qt-project.QtWebKit       	0x0000000111acdf87 0x11099d000 + 18026375
3   org.qt-project.QtWebKit       	0x0000000111ace018 0x11099d000 + 18026520
4   org.qt-project.QtWebKit       	0x0000000111dbc74f 0x11099d000 + 21100367
5   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
6   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
7   libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 8:: Thread (pooled)
0   libsystem_kernel.dylib        	0x00007fff95f0bf5e __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff9098f73d _pthread_cond_wait + 767
2   org.qt-project.QtCore         	0x0000000112e1bbab 0x112de5000 + 224171
3   org.qt-project.QtCore         	0x0000000112e1ba62 QWaitCondition::wait(QMutex*, unsigned long) + 162
4   org.qt-project.QtCore         	0x0000000112e172fe 0x112de5000 + 205566
5   org.qt-project.QtCore         	0x0000000112e1a523 0x112de5000 + 218403
6   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
7   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
8   libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 9:: Thread (pooled)
0   libsystem_kernel.dylib        	0x00007fff95f0bf5e __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff9098f73d _pthread_cond_wait + 767
2   org.qt-project.QtCore         	0x0000000112e1bbab 0x112de5000 + 224171
3   org.qt-project.QtCore         	0x0000000112e1ba62 QWaitCondition::wait(QMutex*, unsigned long) + 162
4   org.qt-project.QtCore         	0x0000000112e172fe 0x112de5000 + 205566
5   org.qt-project.QtCore         	0x0000000112e1a523 0x112de5000 + 218403
6   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
7   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
8   libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 10:: com.apple.NSEventThread
0   libsystem_kernel.dylib        	0x00007fff95f06c96 mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff95f060d7 mach_msg + 55
2   com.apple.CoreFoundation      	0x00007fff8f3472b4 __CFRunLoopServiceMachPort + 212
3   com.apple.CoreFoundation      	0x00007fff8f34677c __CFRunLoopRun + 1356
4   com.apple.CoreFoundation      	0x00007fff8f345fc8 CFRunLoopRunSpecific + 296
5   com.apple.AppKit              	0x00007fff8d702d49 _NSEventThread + 149
6   libsystem_pthread.dylib       	0x00007fff9098e9b1 _pthread_body + 131
7   libsystem_pthread.dylib       	0x00007fff9098e92e _pthread_start + 168
8   libsystem_pthread.dylib       	0x00007fff9098c385 thread_start + 13

Thread 0 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x00007fff50786c08  rcx: 0xffffffffffffffff  rdx: 0x0000000000000000
  rdi: 0x0000000113119e00  rsi: 0x0000000113d63400  rbp: 0x00007fff50786af0  rsp: 0x00007fff50786aa0
   r8: 0x0000000000000007   r9: 0x00007f8fc49d3ba0  r10: 0x00000000a3ff5972  r11: 0x00007f8fc4900000
  r12: 0x00007fff50786c10  r13: 0x0000000000000000  r14: 0x00007f8fc4e36380  r15: 0x0000000000000001
  rip: 0x000000010f6ee834  rfl: 0x0000000000010286  cr2: 0x000000017014a000
  
Logical CPU:     0
Error Code:      0x0100001f
Trap Number:     133
```

**Comment by kevinhendricks** [comment]
Is this repeatable?  Does it happen with our latest release Sigil-0.9.0?  If so, I would be happy to look at it but would need the test case.  The best way to privately send an attachment to me is to post it someplace and then personal mail "KevinH" on the MobileRead Forums (www.mobileread.com) with the link.  Do not post the link here as this is public for the world to see.

Thanks,
KevinH

**Comment by chrisridd** [comment]
I'll drop you a PM.

**Comment by kevinhendricks** [comment]
I just tried it and had no problems.  This is with current master and Mac OS X Yosemite.  So can you please be very specific as to how to create the crash.

The Find setting were:

Dotall, Minimal Match, Wrap
Case-Sensitive
CurrentFile
Searching for "biD"

So 

1. What file were you searching in?
2. Were you in BookView or CodeView mode?
3. Where is the cursor when you began the search?
4. Were you searching Up or Down?

Thanks,

**Comment by kevinhendricks** [comment]
I simply can not recreate this at all with your test case.  But I am using Yosemite.  I have not upgraded to El Capitan due to some concerns about stability.
Perhaps this bug exists in El Capitan only?

If you can provide any more details about how to recreate this bug, please let me know.

**Comment by chrisridd** [comment]
1. Run Sigil.app
2. Double-click the testcase.epub file
3. The validation pane was open, the find pane was open, and the cover page was opened in preview mode.
4.The find field was "biD", the options were DotAll, Minimal Match and Wrap, the Mode was Case Sensitive Current File Down.
5. I tapped the Find button
6. Crash!

I can try trashing the app preferences and retrying - where are they?

**Comment by kevinhendricks** [comment]
So you did not use Sgil to navigate to and open the test epub?  You just opened Sigil and double-clicked the test epub in Finder.  Is that correct?   If so, you must have the .epub extension set to open Sigil.  Mine opens my epub reading software so I have never tried it.

I will play around and see if I can get mine to fail.

The Sigil.ini file is in  your local "Library/Application Support/sigil-ebook/sigil/ " folder someplace.

**Comment by kevinhendricks** [comment]
I was now able to recreate this. Trying to get the cursor on a BookView page with no text (only images) returned a 0.  We now handle this special case.  The fix will appear in the next release.
Thanks for your bug report!
Closing this issue.

**Comment by chrisridd** [comment]
Yes, I have Finder set to open epubs in Sigil.

I just removed the entire ~/Library/Application Support/sigil-ebook/sigil folder (I've a backup in Time Machine), and:

1. Opened Sigil.app
2. Double-clicked the testcase.epub
3. Typed Command-F to bring up the Find pane
4. Entered "biD" and set the same find settings as before.
5. Clicked Find
6. No crash!

So it seems like something in my prefs.. I've restored my old prefs and it looks like the crash only stops after I remove sigil.ini.

Just seen your update - great find! I'll switch to another page before searching next time.
